spinalcordtoolbox/image.py

```diff
@@ -641,14 +648,22 @@ def transfo_pix2phys(self, coordi=None):
             coordi_phys = img.transfo_pix2phys(coordi=coordi_pix)
 
         """
-
-        m_p2f = self.hdr.get_best_affine()
-        aug = np.hstack((np.asarray(coordi), np.ones((len(coordi), 1))))
-        ret = np.empty_like(coordi, dtype=np.float64)
-        for idx_coord, coord in enumerate(aug):
-            phys = np.matmul(m_p2f, coord)
-            ret[idx_coord] = phys[:3]
-        return ret
+        coordi = np.asarray(coordi, dtype=np.float64)
+        num_points, dimension = coordi.shape
+        if dimension != 3:
+            raise ValueError(f'wrong {dimension=}')
+        if mode == 'absolute':
+            affine_column = np.ones((num_points, 1), dtype=np.float64)
+        elif mode == 'relative':
+            affine_column = np.zeros((num_points, 1), dtype=np.float64)
+        else:
+            raise ValueError(f'invalid {mode=}')
+        augmented_pix = np.hstack([coordi, affine_column])
+        # The affine matrix usually transforms _column_ vectors of pix coordinates, but here we
+        # want to transform _row_ vectors of pix coordinates, so we need the transpose.
+        affine_matrix = self.hdr.get_best_affine().T
+        augmented_phys = np.matmul(augmented_pix, affine_matrix)
```

Comment:
On reflection, I kind of like how the old code kept the intuitive "column vector * affine matrix" behavior. It seems a bit strange to transpose the affine matrix, since if we try to introspect the variable, it'll be in a somewhat confusing form. Is there a way we can have our cake and eat it too? (Non-transposed affine matrix, column vectors instead of row vectors.)

Maybe like. Transposing `augmented_pix` to be `[4:N]`, doing the matmul, then transposing it back when returning it?

```suggestion
       # The affine matrix usually transforms _column_ vectors of pix coordinates, but
       # `augmented_pix` takes the form of _row_ vectors. So, we transpose before and 
       # after we do the matrix multiplication.
       affine_matrix = self.hdr.get_best_affine()
       augmented_phys = np.matmul(affine_matrix, augmented_pix.T).T
```

Reply:
Sure, that's nicer. (To be honest, I would have preferred for basically _all_ these vectors to have been column vectors from the start, but that would have been a large and mostly pointless refactor.)

Reply:
Oh no! I'm so sorry, I messed up the indentation level in my suggestion. :facepalm:

Reply:
It's ok, fixed now!